File: cpp/organism/animal/species/Sheep.cpp
//
// Created by Filip on 14/04/2025.
//

#include "Sheep.h"

Sheep::Sheep(Point point, World &world)
    : Animal(point, Constants::Animal::Sheep::Symbol,
             Constants::Animal::Sheep::Strength,
             Constants::Animal::Sheep::Initiative, world) {}

void Sheep::reproduce() {
    Point newPosition = world.getFreeSpace(position);
    if (newPosition == Point(-1, -1)) {
        return;
    }

    world.queueOrganismAddition(new Sheep(newPosition, world));
}

Organism *Sheep::clone() const {
    return new Sheep(*this);// Tworzy kopię obiektu za pomocą konstruktora kopiującego
}

File: java/src/organism/animal/species/Sheep.java
package organism.animal.species;

import constants.Constants;
import logger.EventLogger;
import organism.Organism;
import organism.animal.Animal;
import world.World;

public class Sheep extends Animal {
    public Sheep(int x, int y, World world) {
        super(x, y, Constants.Animal.Sheep.Symbol, Constants.Animal.Sheep.Strength, Constants.Animal.Sheep.Initiative, world);
    }

    @Override
    protected void reproduce() {
        int[] newPosition = world.findNearestFree(this.getX(), this.getY());
        if (newPosition != null) {
            // Create a new organism of the same type
            Animal offspring = new Sheep(newPosition[0], newPosition[1], world);
            world.addOrganism(offspring);
            EventLogger.getInstance().log("Organism " + this.getClass().getSimpleName() + " reproduced at (" + newPosition[0] + ", " + newPosition[1] + ")");
        }
    }

    @Override
    public boolean collision(Organism other) {
        return super.collision(other);
    }
}
